src/bloqade/analysis/fidelity/impls.py

```diff
@@ -0,0 +1,62 @@
+from kirin import interp
+from kirin.analysis import ForwardFrame
+from kirin.dialects import scf
+
+from bloqade.analysis.address import Address
+
+from .analysis import FidelityAnalysis
+
+
+@scf.dialect.register(key="circuit.fidelity")
+class __ScfMethods(interp.MethodTable):
+    @interp.impl(scf.IfElse)
+    def if_else(
+        self, interp_: FidelityAnalysis, frame: ForwardFrame[Address], stmt: scf.IfElse
+    ):
+
+        # NOTE: store a copy of the fidelities
+        current_gate_fidelities = interp_.gate_fidelities
+        current_survival_fidelities = interp_.qubit_survival_fidelities
+
+        # TODO: check if the condition is constant and fix the branch in that case
+        # run both branches
+        with interp_.new_frame(stmt, has_parent_access=True) as then_frame:
+            # NOTE: reset fidelities before stepping into the then-body
+            interp_.reset_fidelities()
+
+            interp_.frame_call_region(
+                then_frame,
+                stmt,
+                stmt.then_body,
+                *(interp_.lattice.bottom() for _ in range(len(stmt.args))),
```

Comment:
two things:
1. IfElse only takes one argument, the condition
2. The block only takes one argument which is the condition.

Because of this you should pass a lattice element with Constant `True` for the then block and `False` for the else block.